### PR TITLE
Fix Content-Type header, should be x-www-form-urlencoded for token request, and not passed for other GET requests

### PR DESCRIPTION
**Pull request by jabbera**
Use correct encoding for token request. 

Fixes: https://github.com/jupyterhub/oauthenticator/issues/598
Fixes: https://github.com/jupyterhub/oauthenticator/issues/564

### CODE REVIEW

**Review by GeorgianaElena — CHANGES_REQUESTED** [review]
Thank you very much for reporting and fixing this @jabbera.

The issue in https://github.com/jupyterhub/oauthenticator/issues/598 seems to be the same with https://github.com/jupyterhub/oauthenticator/issues/564. And since this is impacting all the oauthenticators I believe it's worth modifying the base OAuth class directly instead of allowing it it to be customized by different oautnenticatot classe.

**Comment by jabbera** [comment]
@GeorgianaElena  all set

**Comment by jabbera** [comment]
Does this need anything else?

**Review by consideRatio — COMMENTED** [review]
# Investigation

<details>
<summary>Click to see investigation</summary>

I looked in the previous release of oauthenticator, version 15.1.0, and concluded that Content-Type was set to x-www-form-urlencoded for: azuread, bitbucket, google.

But:
- azuread, bitbucket, google set Content-Type: application/x-www-form-urlencoded with details about utf-8 etc.
- auth0 set Content-Type: application/json
- cilogon, generic, github, gitlab, globus, okpy, openshift provided no Content-Type header
- mediawiki relied on mwoauth library

Since Content-Type influence what is passed, the fix ought to be ensure we declare Content-Type correctly for all requests based on what data we pass.

- auth0, seems like `"Content-Type": "application/json"` is correct
  ```python
  body=json.dumps(params),
  ```
- azread, seems like `'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8'` is correct
  ```python
  data = urllib.parse.urlencode(params, doseq=True, encoding='utf-8', safe='=')
  ```
- bitbucket, seems like `"Content-Type": "application/x-www-form-urlencoded;charset=utf-8"` is correct
  ```python
  body=urllib.parse.urlencode(params).encode('utf-8'),
  ```
- google, seems like `"Content-Type": "application/x-www-form-urlencoded"` is correct
  ```python
  body = urllib.parse.urlencode(
              dict(
                  code=code,
                  redirect_uri=self.get_callback_url(handler),
                  client_id=self.client_id,
                  client_secret=self.client_secret,
                  grant_type="authorization_code",
              )
          )
  ```
- generic, not providing a Content-Type, actually urlencodes params, not setting Content-Type header is likeley incorrect
  ```python
  body=urlencode(params),
  ```
- cilogon, github, gitlab, openshift provides an empty string as POST body - not setting Content-Type header is likely correct
- globus, okpy, makes use of a GET request where there is no POST body - not setting Content-Type header is correct

</details>

# Conclusion

Overall, I think we can't provide a standardized Content-Type across our authenticators, but need to let the authenticators decide on it setting that header or not, and to what.

- For GET requests (globus, okpy) and POST requests with a empty-string body (cilogon, github, gitlab, openshift), we should not set Content-Type at all I think.
- For POST requests where we urlencode params (azuread, google, bitbucket, generic), we should set `application/x-www-form-urlencoded` I think.
- For POST requests where we pass json encoded params (auth0), we should set `application/json` I think.

> **EDIT: New conclusion**
> I got this wrong, I do think we can provide a standarized Content-Type across our authenticators, and that should be what all authenticators support thanks to the OAuth2 specification, x-www-form-urlencoded with an utf-8 charset. But we need to be explicit about the content-type, otherwise some assume json etc.

# Related reading

- [About Content-Type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type)
- [About Content-Type with no body for POST requests](https://stackoverflow.com/a/29784642/2220152)

**Comment by jabbera** [comment]
Thanks. This is the opposite of the other feedback I received which asked me to centralize it instead of doing it at the provider level. I'm happy to revert it and add the other changes you mentioned in the other providers. Are we fairly certain this is the desired end state?

File: oauthenticator/tests/mocks.py [hunk 1]
```diff
@@ -140,30 +140,17 @@ def access_token(request):
...
-            try:
-                body = json.loads(body)
-            except ValueError:
-                return HTTPResponse(
-                    request=request,
-                    code=400,
-                    reason="Body not JSON: %r" % body,
-                )
-            else:
-                code = body['code']
```
Comment (consideRatio) [inline oauthenticator/tests/mocks.py:154]: This section was removed and only influenced tests of auth0 that had been sent JSON in the Post body before, but now sends an urlencoded body like other does.

File: oauthenticator/tests/mocks.py [hunk 2]
```diff
@@ -140,30 +140,17 @@ def access_token(request):
...
+        if not query:
+            query = request.body.decode('utf8')
+        query = parse_qs(query)
+        if 'code' not in query:
+            return HTTPResponse(
+                request=request,
+                code=400,
+                reason=f"No code in access token request: url={request.url}, body={request.body}",
+            )
+        code = query['code'][0]
```
Comment (consideRatio) [inline oauthenticator/tests/mocks.py:153]: This is unchanged code beside intentation change when it was no longer nested in an `else` clause.

File: oauthenticator/tests/test_globus.py [hunk 3]
```diff
@@ -156,7 +156,6 @@ def globus_client(client, mock_globus_token_response):
         access_token_path='/v2/oauth2/token',
         user_path='/v2/oauth2/userinfo',
         token_type='bearer',
-        token_request_style='post',
```
Comment (consideRatio) [inline oauthenticator/tests/test_globus.py:159]: This was the default value anyhow.

File: oauthenticator/tests/test_auth0.py [hunk 4]
```diff
@@ -29,7 +29,6 @@ def auth0_client(client):
         host=auth0_domain,
         access_token_path='/oauth/token',
         user_path='/userinfo',
-        token_request_style='json',
```
Comment (consideRatio) [inline oauthenticator/tests/test_auth0.py:32]: This was no longer accurate.

**Review by consideRatio — APPROVED** [review]
I had to think a lot before feeling confident about this change, the comment https://github.com/jupyterhub/oauthenticator/pull/599#pullrequestreview-1396378578 was a starting point for me to understand this well enough.

I pushed a few commits to finish up some details, but this looks good to me now!

Changes I made:
- Removed irrelevant Content-Type header on GET requests
- Switched to rely on Python's standard library to do the urlencoding
- Updated docstrings with details on Content-Type
- Cleaned up unused test configuration

**Comment by consideRatio** [comment]
Thank you @jabbera for helping us get this fixed! I was wrong to suggest we needed to adjust the Content-Type for individual authenticators I think.

I added an EDIT note in my comment above:

image

**Comment by jabbera** [comment]
Glad we could get this there. The new extension points in the refactor are awesome.

**Review by GeorgianaElena — APPROVED** [review]
Awesome work! Thanks @jabbera, thanks @consideRatio 🎉 🚀

**Comment by jabbera** [comment]
It works!

**Pull request merged**